docs/USAGE-GUIDE.md

```diff
@@ -1777,6 +1777,164 @@ handle. The handle is considered invalid after **s2n_connection_free** is used.
+
+**s2n_psk_set_secret** sets the out-of-band/externally provisioned secret for a given PSK. Note that the secret value is copied into s2n-tls memory and the user is
+responsible for freeing the memory associated with the **secret** input. Be sure to call **free(secret)** after this API call.
+
+**s2n_psk_set_hmac** sets the PSK hash algorithm for a given PSK.
+The supported PSK hash algorithms are listed in the **s2n_psk_hmac** enum. Note that in versions >= TLS1.3 the cipher suite selected has the same Hash alogrithm
+as the PSK. Use this API to override the default PSK Hash Algorithm value of **S2N_PSK_HMAC_SHA256**. From [RFC8446 Section-2.2](https://tools.ietf.org/html/rfc8446#section-2.2) when PSKs are provisioned out-of-band the KDF hash algorithm to be used with the PSK MUST also be provisioned. This API helps setting the KDF hash algorithm to be used with the PSK.
+
+After adding an identity and secret to the PSK object, append the PSK to the connection using **s2n_connection_append_psk**. This API sets up the list of PSKs on the connection that will later be used by the PSK extension to select a PSK for the connection. If a PSK with a duplicate identity is found, an error is returned and the PSK is not added to the list. Note that the PSK object is copied to the connection and the user is responsible for freeing the memory associated with the s2n_psk object. Use **s2n_psk_free** to free the memory associated with the external PSK object after using this API.
+
+**s2n_config_set_psk_mode** sets the PSK mode on the s2n config object. The enum **s2n_psk_mode** lists the supported PSK modes. Use this API prior to the ClientHello to set the PSK mode for the connection. This API can also be used to switch between the different PSK modes.
+
+**s2n_connection_set_psk_mode** sets the PSK mode on the s2n connection object. The supported PSK modes are listed in the enum **s2n_psk_mode** within the **api/s2n.h** file.
+This API overrides the PSK mode for the connection and will continue to use the existing PSK mode even when a new config is set for the connection.
+
+### Obtaining Negotiated Pre-Shared Keys
+
+The following APIs enable the caller to obtain the PSK that was selected by the server for the connection:
+
+```c
+int s2n_connection_get_negotiated_psk_identity_length(struct s2n_connection *conn, uint16_t *identity_length);
+int s2n_connection_get_negotiated_psk_identity(struct s2n_connection *conn, uint8_t *identity, uint16_t max_identity_length);
+```
+
+**s2n_connection_get_negotiated_psk_identity_length** gets the negotiated PSK identity length from the s2n connection object.
+If the negotiated PSK does not exist, the value **0** is returned.
+
+**s2n_connection_get_negotiated_psk_identity** gets the negotiated PSK identity from the s2n connection object.
+If the negotiated PSK does not exist, the PSK identity will not be obtained and no error will be returned.
+
+### Using the Offered Pre-Shared Keys to Select a Pre-Shared Key
```

Comment:
I don't think the "Offered Pre-shared keys" are important enough to make the header. Focus on the actual use case, so something like "Implementing the PSK selection callback" or "Choosing a pre-shared key".

You could also describe the default behavior (choose the first valid resumption psk, choose the first match from the server's list) in this section so that people know if they need to implement the callback.

I also don't see any mention of synchronous / asynchronous. Can this callback be implemented asynchronously?

Reply:
Nope, this callback has to be implemented prior to processing of the psk extensions/ClientHello

Reply:
You should note that then. One of the first questions people are going to ask about this callback is whether it can be called asynchronously.